source/extensions/filters/http/admission_control/admission_control.cc

```diff
@@ -0,0 +1,138 @@
+              ? std::make_unique<Runtime::Double>(proto_config.aggression_coefficient(), runtime_)
+              : nullptr),
+      response_evaluator_(std::move(response_evaluator)) {}
+
+double AdmissionControlFilterConfig::aggression() const {
+  return std::max<double>(1.0, aggression_ ? aggression_->value() : defaultAggression);
+}
+
+AdmissionControlFilter::AdmissionControlFilter(AdmissionControlFilterConfigSharedPtr config,
+                                               const std::string& stats_prefix)
+    : config_(std::move(config)), stats_(generateStats(config_->scope(), stats_prefix)) {}
+
+Http::FilterHeadersStatus AdmissionControlFilter::decodeHeaders(Http::RequestHeaderMap&, bool) {
+  if (!config_->filterEnabled() || decoder_callbacks_->streamInfo().healthCheck()) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  if (shouldRejectRequest()) {
+    decoder_callbacks_->sendLocalReply(Http::Code::ServiceUnavailable, "", nullptr, absl::nullopt,
+                                       "denied by admission control");
+    stats_.rq_rejected_.inc();
+    return Http::FilterHeadersStatus::StopIteration;
+  }
+
+  if (deferred_record_failure_ != nullptr) {
+    // To give a new value to the Cleanup() variable, we must cancel it first to avoid triggering
+    // the function it wraps.
+    deferred_record_failure_->cancel();
+  }
+  deferred_record_failure_ =
+      std::make_unique<Cleanup>([this]() { config_->getController().recordFailure(); });
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterHeadersStatus AdmissionControlFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
+                                                                bool end_stream) {
+  bool successful_response = false;
+  if (Grpc::Common::isGrpcResponseHeaders(headers, end_stream)) {
+    absl::optional<GrpcStatus> grpc_status = Grpc::Common::getGrpcStatus(headers);
+
+    // If the GRPC status isn't found in the headers, it must be found in the trailers.
+    expect_grpc_status_in_trailer_ = !grpc_status.has_value();
+    if (expect_grpc_status_in_trailer_) {
+      return Http::FilterHeadersStatus::Continue;
+    }
+
+    const uint32_t status = enumToInt(grpc_status.value());
+    successful_response = config_->responseEvalutor().isGrpcSuccess(status);
+  } else {
+    // HTTP response.
+    const uint64_t http_status = Http::Utility::getResponseStatus(headers);
+    successful_response = config_->responseEvalutor().isHttpSuccess(http_status);
+  }
+
+  if (successful_response) {
+    recordSuccess();
+  } else {
+    recordFailure();
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterTrailersStatus
+AdmissionControlFilter::encodeTrailers(Http::ResponseTrailerMap& trailers) {
+  if (expect_grpc_status_in_trailer_) {
+    absl::optional<GrpcStatus> grpc_status = Grpc::Common::getGrpcStatus(trailers, false);
+
+    // Status code must be sent in trailers.
+    ASSERT(grpc_status.has_value());
```

Comment:
No guarantee the upstream does this. Don't use ASSERT here and handle errors (and test).

Reply:
My reason for this was that I read this part in the [gRPC over HTTP2 protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) (emphasis mine):

> Response-Headers & Trailers-Only are each delivered in a single HTTP2 HEADERS frame block. Most responses are expected to have both headers and trailers but Trailers-Only is permitted for calls that produce an immediate error. **Status must be sent in Trailers even if the status code is OK**.

I interpreted that to mean that we can always expect a status. This should definitely be handled differently, I'm just unsure of what to do in this case- maybe return a `INVALID_ARGUMENT` status?

Reply:
I would synthesize a failure value and just increment it (or something like that). It doesn't really matter.